zig cannot build against a native glibc greater than the latest cross-compilation-supported version of glibc

### Zig Version

none?

### Steps to Reproduce

1) Try to build Zig on systems with glibc 2.35 and newer

Related: https://github.com/ziglang/zig/pull/12567#pullrequestreview-1089121587
> This change does mean linking libc will be broken on OS's with too-new libc (e.g. NixOS and Arch) until we have a fallback to link against system libc or update Zig-provided glibc, but this still seems like a step in the right direction even without that

And https://github.com/ziglang/zig/pull/12788#issuecomment-1241768153
> Can confirm that this detects the correct glibc version for me and I can build the example in https://github.com/ziglang/zig/issues/11137 **if I explicitly specify an earlier version that Zig can provide.**

### Expected Behavior

Zig stage3 builded succesfully.

### Actual Behavior

```sh
[195/195] cd /tmp/portage/dev-lang/zig-9999/work/zig-9999 && /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/zig2 build --zig-lib-dir /tmp/portage/dev-lang/zig-9999/work/zig-9999/lib --prefix /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/stage3 -Dconfig_h=/tmp/portage/dev-lang/zig-9999/work/zig-9999/build/config.h -Denable-llvm -Denable-stage1 -Drelease -Dskip-install-lib-files=false -Dtarget=native -Dcpu=baseline
FAILED: CMakeFiles/stage3 /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/CMakeFiles/stage3 
cd /tmp/portage/dev-lang/zig-9999/work/zig-9999 && /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/zig2 build --zig-lib-dir /tmp/portage/dev-lang/zig-9999/work/zig-9999/lib --prefix /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/stage3 -Dconfig_h=/tmp/portage/dev-lang/zig-9999/work/zig-9999/build/config.h -Denable-llvm -Denable-stage1 -Drelease -Dskip-install-lib-files=false -Dtarget=native -Dcpu=baseline
error: zig does not yet provide glibc version 2.35, the max provided version is 2.34
error: unable to build glibc shared objects: InvalidTargetGLibCVersion
error: zig...
error: The following command exited with error code 1:
/tmp/portage/dev-lang/zig-9999/work/zig-9999/build/zig2 build-exe --stack 33554432 /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1.zig -lc -cflags -std=c++14 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D_GNU_SOURCE -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -Werror=type-limits -Wno-missing-braces -Wno-comment -- /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/analyze.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/astgen.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/bigfloat.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/bigint.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/buffer.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/codegen.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/errmsg.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/error.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/heap.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/ir.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/ir_print.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/mem.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/os.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/parser.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/range_set.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/stage1.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/target.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/tokenizer.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/util.cpp /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/softfloat_ext.cpp -cflags -std=c99 -O3 -- /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/stage1/parse_f128.c /tmp/portage/dev-lang/zig-9999/work/zig-9999/zig-cache/o/866f9b3a9b210d872d324ab066f29d86/libsoftfloat.a -lc++ /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/zigcpp/libzigcpp.a /usr/lib/llvm/14/lib64/libclang-cpp.so.14 /usr/lib64/liblldMinGW.so /usr/lib64/liblldELF.so /usr/lib64/liblldCOFF.so /usr/lib64/liblldWasm.so /usr/lib64/liblldMachO.so /usr/lib64/liblldCommon.so /usr/lib/llvm/14/lib64/libLLVM-14.so /usr/lib/gcc/x86_64-pc-linux-gnu/11.3.0/libstdc++.so -lunwind -OReleaseFast --cache-dir /tmp/portage/dev-lang/zig-9999/work/zig-9999/zig-cache --global-cache-dir /tmp/portage/dev-lang/zig-9999/homedir/.cache/zig --name zig -target native-native -mcpu x86_64 --pkg-begin build_options /tmp/portage/dev-lang/zig-9999/work/zig-9999/zig-cache/options/xgA7uMtMiAJk0oy41EHIvQ8B1fl4rxmMBq3gbl2NVySjDFlN626oCTSRj32_6Ae5 --pkg-end --pkg-begin compiler_rt /tmp/portage/dev-lang/zig-9999/work/zig-9999/src/empty.zig --pkg-end -I /tmp/portage/dev-lang/zig-9999/work/zig-9999/src -I /tmp/portage/dev-lang/zig-9999/work/zig-9999/deps/SoftFloat-3e/source/include -I /tmp/portage/dev-lang/zig-9999/work/zig-9999/deps/SoftFloat-3e-prebuilt -I /usr/include -D ZIG_LINK_MODE=Static -L /usr/lib/llvm/14/lib -isystem /usr/lib/llvm/14/include -fno-build-id --zig-lib-dir /tmp/portage/dev-lang/zig-9999/work/zig-9999/lib --enable-cache 
error: the following build command failed with exit code 1:
/tmp/portage/dev-lang/zig-9999/work/zig-9999/zig-cache/o/73e5f8c2a8fd1c8c2a10413a1c609d5e/build /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/zig2 /tmp/portage/dev-lang/zig-9999/work/zig-9999 /tmp/portage/dev-lang/zig-9999/work/zig-9999/zig-cache /tmp/portage/dev-lang/zig-9999/homedir/.cache/zig --zig-lib-dir /tmp/portage/dev-lang/zig-9999/work/zig-9999/lib --prefix /tmp/portage/dev-lang/zig-9999/work/zig-9999/build/stage3 -Dconfig_h=/tmp/portage/dev-lang/zig-9999/work/zig-9999/build/config.h -Denable-llvm -Denable-stage1 -Drelease -Dskip-install-lib-files=false -Dtarget=native -Dcpu=baseline
ninja: build stopped: subcommand failed.
```

I hit this on Ubuntu 22.04.01 aarch64 trying to build a trivial C program, i.e., this isn't just a problem building zig itself, but a problem building any C program at all:

```sh
~/python-wasm/packages/lzma/build/native$ cat /etc/issue
Ubuntu 22.04.1 LTS \n \l
~/python-wasm/packages/lzma/build/native$ zig version
0.10.0-dev.3978+4fd4c733d
~/python-wasm/packages/lzma/build/native$ more a.c
main() { printf("hi\n");}
~/python-wasm/packages/lzma/build/native$ zig cc a.c
error: zig does not yet provide glibc version 2.35, the max provided version is 2.34
error: unable to build glibc shared objects: InvalidTargetGLibCVersion
~/python-wasm/packages/lzma/build/native$ zig cc -target x86_64-musl a.c
error: UnknownOperatingSystem
~/python-wasm/packages/lzma/build/native$ zig cc -target aarch64-linux-musl a.c
LLD Link... a.c:1:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
main() { printf("hi\n");}
^
a.c:1:10: warning: implicitly declaring library function 'printf' with type 'int (const char *, ...)' [-Wimplicit-function-declaration]
main() { printf("hi\n");}
         ^
a.c:1:10: note: include the header <stdio.h> or explicitly provide a declaration for 'printf'
2 warnings generated.
~/python-wasm/packages/lzma/build/native$ ./a.out 
hi
~/python-wasm/packages/lzma/build/native$ zig cc -target aarch64-linux-gnu.2.34 a.c
a.c:1:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
main() { printf("hi\n");}
^
a.c:1:10: warning: implicitly declaring library function 'printf' with type 'int (const char *, ...)' [-Wimplicit-function-declaration]
main() { printf("hi\n");}
         ^
a.c:1:10: note: include the header <stdio.h> or explicitly provide a declaration for 'printf'
2 warnings generated.
~/python-wasm/packages/lzma/build/native$ ./a.out 
hi
}}}

--- a/lib/std/child_process.zig
+++ b/lib/std/child_process.zig
@@ -508,7 +508,7 @@ pub const ChildProcess = struct {
                 // it, that's the error code returned by the child process.
                 _ = std.os.poll(&fd, 0) catch unreachable;
 
-                // According to eventfd(2) the descriptro is readable if the counter
+                // According to eventfd(2) the descriptor is readable if the counter
                 // has a value greater than 0
                 if ((fd[0].revents & std.os.POLL.IN) != 0) {
                     const err_int = try readIntFd(err_pipe[0]);

--- a/src/Compilation.zig
+++ b/src/Compilation.zig
@@ -4549,74 +4549,29 @@ fn detectLibCIncludeDirs(
                 getZigShippedLibCIncludeDirsDarwin(arena, zig_lib_dir, target);
         }
         const libc = try arena.create(LibCInstallation);
-        libc.* = try LibCInstallation.findNative(.{ .allocator = arena, .verbose = true });
+        libc.* = LibCInstallation.findNative(.{ .allocator = arena }) catch |err| switch (err) {
+            error.CCompilerExitCode,
+            error.CCompilerCrashed,
+            error.CCompilerCannotFindHeaders,
+            error.UnableToSpawnCCompiler,
+            => |e| {
+                // We tried to integrate with the native system C compiler,
+                // however, it is not installed. So we must rely on our bundled
+                // libc files.
+                if (target_util.canBuildLibC(target)) {
+                    return detectLibCFromBuilding(arena, zig_lib_dir, target, has_macos_sdk);
+                }
+                return e;
+            },
+            else => |e| return e,
+        };
         return detectLibCFromLibCInstallation(arena, target, libc);
     }
 
     // If not linking system libraries, build and provide our own libc by
     // default if possible.
     if (target_util.canBuildLibC(target)) {
-        switch (target.os.tag) {
-            .macos => return if (has_macos_sdk)
-                // For Darwin/macOS, we are all set with getDarwinSDK found earlier.
-                LibCDirs{
-                    .libc_include_dir_list = &[0][]u8{},
-                    .libc_installation = null,
-                }
-            else
-                getZigShippedLibCIncludeDirsDarwin(arena, zig_lib_dir, target),
-            else => {
-                const generic_name = target_util.libCGenericName(target);
-                // Some architectures are handled by the same set of headers.
-                const arch_name = if (target.abi.isMusl())
-                    musl.archName(target.cpu.arch)
-                else if (target.cpu.arch.isThumb())
-                    // ARM headers are valid for Thumb too.
-                    switch (target.cpu.arch) {
-                        .thumb => "arm",
-                        .thumbeb => "armeb",
-                        else => unreachable,
-                    }
-                else
-                    @tagName(target.cpu.arch);
-                const os_name = @tagName(target.os.tag);
-                // Musl's headers are ABI-agnostic and so they all have the "musl" ABI name.
-                const abi_name = if (target.abi.isMusl()) "musl" else @tagName(target.abi);
-                const s = std.fs.path.sep_str;
-                const arch_include_dir = try std.fmt.allocPrint(
-                    arena,
-                    "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "{s}-{s}-{s}",
-                    .{ zig_lib_dir, arch_name, os_name, abi_name },
-                );
-                const generic_include_dir = try std.fmt.allocPrint(
-                    arena,
-                    "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "generic-{s}",
-                    .{ zig_lib_dir, generic_name },
-                );
-                const generic_arch_name = target_util.osArchName(target);
-                const arch_os_include_dir = try std.fmt.allocPrint(
-                    arena,
-                    "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "{s}-{s}-any",
-                    .{ zig_lib_dir, generic_arch_name, os_name },
-                );
-                const generic_os_include_dir = try std.fmt.allocPrint(
-                    arena,
-                    "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "any-{s}-any",
-                    .{ zig_lib_dir, os_name },
-                );
-
-                const list = try arena.alloc([]const u8, 4);
-                list[0] = arch_include_dir;
-                list[1] = generic_include_dir;
-                list[2] = arch_os_include_dir;
-                list[3] = generic_os_include_dir;
-
-                return LibCDirs{
-                    .libc_include_dir_list = list,
-                    .libc_installation = null,
-                };
-            },
-        }
+        return detectLibCFromBuilding(arena, zig_lib_dir, target, has_macos_sdk);
     }
 
     // If zig can't build the libc for the target and we are targeting the
@@ -4673,6 +4628,75 @@ fn detectLibCFromLibCInstallation(arena: Allocator, target: Target, lci: *const 
         .libc_include_dir_list = list.items,
         .libc_installation = lci,
     };
+}
+
+fn detectLibCFromBuilding(
+    arena: Allocator,
+    zig_lib_dir: []const u8,
+    target: std.Target,
+    has_macos_sdk: bool,
+) !LibCDirs {
+    switch (target.os.tag) {
+        .macos => return if (has_macos_sdk)
+            // For Darwin/macOS, we are all set with getDarwinSDK found earlier.
+            LibCDirs{
+                .libc_include_dir_list = &[0][]u8{},
+                .libc_installation = null,
+            }
+        else
+            getZigShippedLibCIncludeDirsDarwin(arena, zig_lib_dir, target),
+        else => {
+            const generic_name = target_util.libCGenericName(target);
+            // Some architectures are handled by the same set of headers.
+            const arch_name = if (target.abi.isMusl())
+                musl.archName(target.cpu.arch)
+            else if (target.cpu.arch.isThumb())
+                // ARM headers are valid for Thumb too.
+                switch (target.cpu.arch) {
+                    .thumb => "arm",
+                    .thumbeb => "armeb",
+                    else => unreachable,
+                }
+            else
+                @tagName(target.cpu.arch);
+            const os_name = @tagName(target.os.tag);
+            // Musl's headers are ABI-agnostic and so they all have the "musl" ABI name.
+            const abi_name = if (target.abi.isMusl()) "musl" else @tagName(target.abi);
+            const s = std.fs.path.sep_str;
+            const arch_include_dir = try std.fmt.allocPrint(
+                arena,
+                "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "{s}-{s}-{s}",
+                .{ zig_lib_dir, arch_name, os_name, abi_name },
+            );
+            const generic_include_dir = try std.fmt.allocPrint(
+                arena,
+                "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "generic-{s}",
+                .{ zig_lib_dir, generic_name },
+            );
+            const generic_arch_name = target_util.osArchName(target);
+            const arch_os_include_dir = try std.fmt.allocPrint(
+                arena,
+                "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "{s}-{s}-any",
+                .{ zig_lib_dir, generic_arch_name, os_name },
+            );
+            const generic_os_include_dir = try std.fmt.allocPrint(
+                arena,
+                "{s}" ++ s ++ "libc" ++ s ++ "include" ++ s ++ "any-{s}-any",
+                .{ zig_lib_dir, os_name },
+            );
+
+            const list = try arena.alloc([]const u8, 4);
+            list[0] = arch_include_dir;
+            list[1] = generic_include_dir;
+            list[2] = arch_os_include_dir;
+            list[3] = generic_os_include_dir;
+
+            return LibCDirs{
+                .libc_include_dir_list = list,
+                .libc_installation = null,
+            };
+        },
+    }
 }
 
 pub fn get_libc_crt_file(comp: *Compilation, arena: Allocator, basename: []const u8) ![]const u8 {

--- a/src/Compilation.zig
+++ b/src/Compilation.zig
@@ -1238,7 +1238,6 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
             options.target,
             options.is_native_abi,
             link_libc,
-            options.system_lib_names.len != 0 or options.frameworks.count() != 0,
             options.libc_installation,
             options.native_darwin_sdk != null,
         );
@@ -4522,7 +4521,6 @@ fn detectLibCIncludeDirs(
     target: Target,
     is_native_abi: bool,
     link_libc: bool,
-    link_system_libs: bool,
     libc_installation: ?*const LibCInstallation,
     has_macos_sdk: bool,
 ) !LibCDirs {
@@ -4539,7 +4537,7 @@ fn detectLibCIncludeDirs(
 
     // If linking system libraries and targeting the native abi, default to
     // using the system libc installation.
-    if (link_system_libs and is_native_abi and !target.isMinGW()) {
+    if (is_native_abi and !target.isMinGW()) {
         if (target.isDarwin()) {
             return if (has_macos_sdk)
                 // For Darwin/macOS, we are all set with getDarwinSDK found earlier.

--- a/src/glibc.zig
+++ b/src/glibc.zig
@@ -719,17 +719,16 @@ pub fn buildSharedObjects(comp: *Compilation) !void {
                 .lt => continue,
                 .gt => {
                     // TODO Expose via compile error mechanism instead of log.
-                    log.err("invalid target glibc version: {}", .{target_version});
+                    log.warn("invalid target glibc version: {}", .{target_version});
                     return error.InvalidTargetGLibCVersion;
                 },
             }
-        } else {
+        } else blk: {
             const latest_index = metadata.all_versions.len - 1;
-            // TODO Expose via compile error mechanism instead of log.
-            log.err("zig does not yet provide glibc version {}, the max provided version is {}", .{
+            log.warn("zig cannot build new glibc version {}; providing instead {}", .{
                 target_version, metadata.all_versions[latest_index],
             });
-            return error.InvalidTargetGLibCVersion;
+            break :blk latest_index;
         };
 
         {

